index.js

```diff
@@ -1,11 +1,264 @@
+        type: 'select',
+        choices: [
+          {
+            label: 'Single Day',
+            value: 'single'
+          },
+          {
+            label: 'Consecutive Days',
+            value: 'consecutive'
+          },
+          {
+            label: 'Recurring',
+            value: 'repeat'
+          }
+        ],
+        def: 'single'
+      },
+      endDate: {
+        label: 'End date',
+        type: 'date',
+        if: {
+          dateType: 'consecutive'
+        }
+      },
+      repeatInterval: {
+        label: 'How often does the event repeat?',
+        type: 'select',
+        choices: [
+          {
+            label: 'Every week',
+            value: 'weeks'
+          },
+          {
+            label: 'Every month',
+            value: 'months'
+          }
+        ],
+        if: {
+          dateType: 'repeat'
+        }
+      },
+      repeatCount: {
+        label: 'How many times does it repeat?',
+        type: 'integer',
+        def: 1,
+        if: {
+          dateType: 'repeat'
+        }
+      },
+      description: {
+        type: 'string',
+        label: 'Description',
+        textarea: true,
+        required: true
+      }
+    },
+    group: {
+      basics: {
+        label: 'Basics',
+        fields: [
+          'title',
+          'slug',
+          'description',
+          'startDate',
+          'allDay',
+          'startTime',
+          'endTime'
+        ]
+      },
+      advanced: {
+        label: 'Advanced',
+        fields: ['dateType', 'endDate', 'repeatInterval', 'repeatCount']
+      },
+      meta: {
+        label: 'Meta',
+        fields: ['tags', 'published']
+      }
+    }
+  },
+  handlers(self, options) {
+    return {
+      beforeSave: {
+        async denormalizeDatesAndTimes(req, piece, options) {
+          self.denormalizeDatesAndTimes(piece);
+        }
+      },
+      beforeInsert: {
+        setGroupId(req, piece, options) {
+          // Set groupId on parent if this is a repeating item
+          if (
+            piece.dateType === 'repeat' &&
+            !piece.groupId &&
+            piece.aposMode === 'draft'
+          ) {
+            piece.groupId = self.apos.util.generateId();
+          }
+        }
+      },
+      afterInsert: {
+        async createRepeatItems(req, piece, options) {
+          if (piece.aposMode === 'draft') {
+            // Workflow is replicating this but also its existing
+            // scheduled repetitions, don't re-replicate them and cause problems
+            return;
+          }
+          if (piece.dateType === 'repeat' && piece.aposMode === 'draft') {
+            await self.repeatEvent(req, piece, options);
+          }
+        }
+      },
+      afterPublish: {
+        async publishChildren(req, piece, options) {
+          // If this is a repeating item, publish its children also
+          if (piece.published.dateType === 'repeat' && piece.firstTime) {
```

Comment:
I'm down to just one thing: where is this `firstTime` property coming from? I think this is the only mention of it in the PR.

Reply:
Maybe I misunderstood - you made a recommendation earlier in the life of this PR to use firstTime to make sure that we were only creating repeat items on first publishing the event.  I assumed firstTime was a property of the piece type that we're extending here?

"The afterPublished event could be used to handle this. If the event that was just published has dateType: repeat and firstTime is true, its repetitions should also be published at that time."

Reply:
Thanks for the context, now I know the answer. It's in `options.firstTime` in this event handler. Not `piece.firstTime`. So if you make that change we should be good to merge.

Reply:
OK - I tried a few different things tonight.  If I switch to options.firstTime, then I get a "cannot read property firstTime of undefined".  Dumping the piece part to the console, there is a firstTime property on the piece object and this seems  to work as I would expect...

I did have to update the createRepeatItems method in testing - as it was did not create any repeating items.

Hope this all works for you to merge, but let me know if there's more to adjust.